packages/xo-web/src/xo-app/sr/tab-general.js

```diff
@@ -2,39 +2,113 @@ import _ from 'intl'
 import { addTag, removeTag, getLicense } from 'xo'
 import { connectStore, formatSize } from 'utils'
 import { Container, Row, Col } from 'grid'
-import { createGetObject } from 'selectors'
+import { createGetObject, createSelector } from 'selectors'
+import {
+  differenceBy,
+  flatMap,
+  forEach,
+  groupBy,
+  identity,
+  keyBy,
+  sumBy,
+} from 'lodash'
+import { generateId } from 'reaclette-utils'
 import { renderXoItemFromId } from 'render-xo-item'
 
 const UsageTooltip = connectStore(() => ({
-  vbd: createGetObject((_, { vdi }) => vdi.$VBDs[0]),
-}))(({ vbd, vdi }) => (
-  <span>
-    {vdi.name_label} − {formatSize(vdi.usage)}
-    {vbd != null && <br />}
-    {vbd != null && renderXoItemFromId(vbd.VM)}
-  </span>
-))
+  vbd: createGetObject((_, { vdi }) => vdi.$VBDs && vdi.$VBDs[0]),
+}))(
+  ({ vbd, vdi: { baseCopyOf, name_label: nameLabel, snapshotOf, usage } }) => {
+    const vdiOf =
+      baseCopyOf === undefined
+        ? snapshotOf === nameLabel || snapshotOf === undefined
+          ? undefined
+          : snapshotOf
+        : baseCopyOf
+    return (
+      <span>
+        {vdiOf === undefined
+          ? `${nameLabel} - ${formatSize(usage)}`
+          : `${nameLabel} of ${vdiOf} - ${formatSize(usage)}`}
+        {vbd != null && <br />}
+        {vbd != null && renderXoItemFromId(vbd.VM)}
+      </span>
+    )
+  }
+)
 
 export default class TabGeneral extends Component {
   componentDidMount() {
     const { sr } = this.props
-
     if (sr.SR_type === 'xosan') {
       getLicense('xosan.trial', sr.id).then(() =>
         this.setState({ licenseRestriction: true })
       )
     }
   }
 
-  render() {
-    const { sr, vdis, vdiSnapshots, unmanagedVdis } = this.props
+  _getVdis = createSelector(
+    () => this.props.vdis,
+    () => this.props.vdiSnapshots,
+    () => this.props.unmanagedVdis,
+    (vdis, vdiSnapshots, unmanagedVdis) => {
+      const spanpshotsByVdis = groupBy(vdiSnapshots, '$snapshot_of')
+      const unmanagedVdisById = keyBy(unmanagedVdis, 'id')
+      let disks = []
+      let bc
+      let snapshots
+      forEach(vdis, vdi => {
+        const { id: vdiId, name_label: vdiNameLabel, smConfig } = vdi
+        bc = unmanagedVdisById[smConfig['vhd-parent']]
+        while (bc !== undefined) {
+          disks.push({ ...bc, baseCopyOf: vdiNameLabel })
+          bc = unmanagedVdisById[bc.smConfig['vhd-parent']]
+        }
+        disks.push(vdi)
+        if ((snapshots = spanpshotsByVdis[vdiId]) !== undefined) {
+          disks.push(
+            snapshots.map(snapshot => ({
+              ...snapshot,
+              snapshotOf: vdiNameLabel,
+            }))
+          )
+        }
+      })
+
+      if ((snapshots = spanpshotsByVdis[undefined]) !== undefined) {
+        disks.unshift(snapshots)
+      }
+      disks = flatMap(disks, identity)
 
+      const diff = differenceBy(
+        unmanagedVdis,
+        disks.filter(({ type }) => type === 'VDI-unmanaged'),
+        'id'
+      )
```

Comment:
Maybe you could remove them from a set while you're going through them the first time to avoid running another loop.